File: 基礎題庫/Contents/c184/c184.cpp
#include<iostream>

using namespace std;

int isPerfect(int);

int main(void)
{
	int n, temp;
	
	while(1)
	{
		cin >> n;
		
		if(!n)
			break;
			
		temp = isPerfect(n);
		if(n == temp)
			cout << "=" << n << endl;
		else
			cout << (isPerfect(temp) == n ? temp : 0) << endl;
	}
	
	return 0;
}

int isPerfect(int num)
{
	int sum = 1;
	
	for(int i = 2; i < num; i++)
		if(!(num % i))
			sum += i;
	return sum;
}

File: 基礎題庫/Contents/c184/c184.py
import sys

def isPerfect(num):
    end = int(num ** 0.5)
    res = end ** 2

    if res == num:
        res = 1 - end
    else:
        res = 1
    
    for i in range(2, end + 1):
        if num % i == 0:
            res += i + num // i
    return res

for n in sys.stdin:
    n = int(n)

    if n == 0:
        break
    
    temp = isPerfect(n)
    if temp == n:
        print("=%d" % n)
    else:
        if isPerfect(temp) == n:
            print(temp)
        else:
            print(0)
